### PR TITLE
Implement terrain transitions in Editor.

### PR DESCRIPTION
**Pull request by Districh-ru**
Relates to https://github.com/ihhub/fheroes2/issues/6845

This PR implements connection of nearby terrains by applying appropriate tile images.

https://github.com/ihhub/fheroes2/assets/113276641/3ce9775e-7477-4108-8810-dc27bdf2895b

There could be "incorrect" tiles when they cannot be properly connected (with transition) with the nearby terrains. This issue probably needs more polishing (which possibly should be done in one of the next PRs).

https://github.com/ihhub/fheroes2/assets/113276641/4430d055-37f2-4731-a640-9e3061b72ac7

### CODE REVIEW

File: src/fheroes2/maps/maps_tiles_helper.cpp [hunk 1]
```diff
@@ -229,36 +231,592 @@ namespace
...
+    }
+
+    // Returns true if terrain transition on tile was properly set or it is not needed.
+    bool setTerrainBoundaries( const int groundDirection, const int beachDirection, const int32_t tileId, const uint16_t imageOffset )
+    {
+        if ( groundDirection == DIRECTION_ALL ) {
+            // No transition is needed.
+            return true;
+        }
+        if ( groundDirection == ( Direction::TOP_RIGHT | Direction::TOP | DIRECTION_BOTTOM_ROW | DIRECTION_CENTER_ROW ) ) {
```
Comment (ihhub) [inline src/fheroes2/maps/maps_tiles_helper.cpp:279]: For better code readability I suggest to cache **world.GetTiles( tileId )** just after the first if-body.

File: src/fheroes2/maps/maps_tiles_helper.cpp [hunk 2]
```diff
@@ -229,36 +231,592 @@ namespace
...
+        return false;
+    }
+
+    // Returns true if terrain transition was set or it is not needed.
+    bool updateTerrainTransitionOnTile( const int32_t tileId )
+    {
+        const int ground = world.GetTiles( tileId ).GetGround();
+
+        if ( ground == Maps::Ground::BEACH ) {
+            // Beach tile images does not have transition with the other terrains.
```
Comment (ihhub) [inline src/fheroes2/maps/maps_tiles_helper.cpp:602]: images **do** - plural form.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @Districh-ru , I left several comments here. Would you mind please to check them?

File: src/fheroes2/maps/ground.cpp [hunk 3]
```diff
@@ -30,6 +30,92 @@
 #include "skill.h"
 #include "translations.h"
 
+uint16_t Maps::Ground::getTerrainIageOffset( const int groundId )
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/maps/ground.cpp:33]: Hi @ihhub, thanks!
I thought about such renaming and implementing `getTerrainEndImageIndex()`, but later decided not to introduce this new function.

File: src/fheroes2/maps/maps.cpp [hunk 4]
```diff
@@ -193,6 +193,35 @@ int32_t Maps::GetDirectionIndex( int32_t from, int vector )
...
+    case Direction::BOTTOM_LEFT:
+        return { from.x - 1, from.y + 1 };
+    case Direction::LEFT:
+        return { from.x - 1, from.y };
+    case Direction::TOP_LEFT:
+        return { from.x - 1, from.y - 1 };
+    case Direction::CENTER:
+        return from;
+    default:
+        break;
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/maps/maps.cpp:218]: Yes, this is not a valid case, in this case function returns {-1,0} (like the upper one function returns index = -1). But for debugging purposes it is better to add an assertion. :)

File: src/fheroes2/maps/maps.cpp [hunk 5]
```diff
@@ -193,6 +193,35 @@ int32_t Maps::GetDirectionIndex( int32_t from, int vector )
     return -1;
 }
 
+fheroes2::Point Maps::getDirectionPoint( const fheroes2::Point & from, int vector )
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/maps/maps.cpp:196]: Sure.
I'll also update this name in the upper function.

File: src/fheroes2/maps/maps_tiles_helper.cpp [hunk 6]
```diff
@@ -229,36 +231,592 @@ namespace
...
+
+            if ( world.GetTiles( tilePosition.x, tilePosition.y ).GetGround() == groundId ) {
+                groundDirection |= direction;
+            }
+        }
+
+        return groundDirection;
+    }
+
+    inline bool hasBits( const int value, const int bits )
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/maps/maps_tiles_helper.cpp:262]: Sure, let's leave this optimization to the compilers. :)

File: src/fheroes2/maps/maps_tiles_helper.cpp [hunk 7]
```diff
@@ -229,36 +231,592 @@ namespace
...
+            // Center tile is out of the map borders.
+            return Direction::UNKNOWN;
+        }
+
+        const fheroes2::Point centerTile = Maps::GetPoint( centerTileIndex );
+        const fheroes2::Point lastWorldTile = { world.w() - 1, world.h() - 1 };
+
+        int groundDirection = ( world.GetTiles( centerTileIndex ).GetGround() == groundId ) ? Direction::CENTER : 0;
+
+        for ( const int & direction : Direction::All() ) {
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/maps/maps_tiles_helper.cpp:248]: `getAroundIndexes()` skips indexes that are out or map boundaries while here we need to treat the nearby indexes for this case to get proper ground transitions on map edges.
Also we need to accumulate the directions while `getAroundIndexes()` returns indexes that should be converted into directions.

File: src/fheroes2/maps/maps_tiles_helper.cpp (hunk 1 above)
Comment (Districh-ru) [reply src/fheroes2/maps/maps_tiles_helper.cpp:279]: Sure)

File: src/fheroes2/maps/maps_tiles_helper.cpp (hunk 2 above)
Comment (Districh-ru) [reply src/fheroes2/maps/maps_tiles_helper.cpp:602]: Oh, sorry...

File: src/fheroes2/maps/maps_tiles_helper.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/maps/maps_tiles_helper.cpp:248]: But I'll use it in `updateTerrainTransitionOnArea()`.
Thanks for the suggestion!

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
Hi @drevoborod , can you please help to test these changes?

**Comment by Branikolog** [comment]
Hi, @Districh-ru !
I've tested this PR and find several issues.
Overall, I like how editor works now. You've done a great job!

https://github.com/ihhub/fheroes2/assets/55348946/4eba39cd-cea8-49aa-a267-d816b29715b9

https://github.com/ihhub/fheroes2/assets/55348946/1fe236fa-7375-4c5d-8fdc-d64191030289

https://github.com/ihhub/fheroes2/assets/55348946/f4ee47aa-87a3-4cd0-b09a-35d970c6fd01

https://github.com/ihhub/fheroes2/assets/55348946/32cfaf7c-8afb-4d91-a593-a2e72613cad2

https://github.com/ihhub/fheroes2/assets/55348946/ec1e50a6-5638-4919-ba04-ff4b62681662

https://github.com/ihhub/fheroes2/assets/55348946/152a2f47-6588-478a-be5f-d8596b50253e

https://github.com/ihhub/fheroes2/assets/55348946/57b80a94-6683-4dab-8b62-c33bc330a420

https://github.com/ihhub/fheroes2/assets/55348946/378849b8-7885-4bdf-b7f5-bca25928a30d

**Comment by Branikolog** [comment]
The smallest brush is not working properly. I know, it will work better, when drag-to-paint will be implemented, but for now it sometimes creates a terrain, which is not selected, like water.

https://github.com/ihhub/fheroes2/assets/55348946/551b2c51-0279-4ddd-8b06-114c64d6e16d

**Comment by Districh-ru** [comment]
Hi, @Branikolog, thanks for testing and for videos!
I'll try to fix most of the found issues in the next days.

**Comment by Districh-ru** [comment]
Fixed issues on videos 5 and 9. Implemented behavior like in HotA map editor as it looks better. :)
<details><summary>Screenshot</summary>

изображение
</details>

**Comment by Districh-ru** [comment]
Hi, @Branikolog, I tried to make a temporary patch for mentioned issues.
IMHO we should return to their fix and redo the terrain connecting algorithm after we implement the UNDO function and after we design new terrain transition images (using the existing ones).

**Review by Branikolog — APPROVED** [review]
After testing the latest build I can tell that everything works well now. Great job, @Districh-ru !

**Comment by ihhub** [comment]
@Districh-ru , huge thanks for these new updates!

**Pull request merged**